### PR TITLE
punishproducer functionality added

### PR DESCRIPTION
**Pull request by Art17**
New action punishprod

eosio_global_rem_state
      microseconds producer_max_inactivity_time = eosio::minutes(30);
      microseconds producer_inactivity_punishment_period = eosio::days(30);

producer_info
     time_point last_block_time
     time_point top21_chosen_time

### CODE REVIEW

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 1]
```diff
@@ -182,6 +182,30 @@ namespace eosiosystem {
          _gstate.current_round_start_time = timestamp;
          _gstate.last_schedule_version = schedule_version;
          std::vector<name> active_producers = eosio::get_active_producers();
+
+         for (size_t producer_index = 0; producer_index < _gstate.last_schedule.size(); producer_index++) {
```
Comment (roman-tik) [inline contracts/contracts/rem.system/src/producer_pay.cpp]: do we need a separate `for` here? maybe join it with the previous one?

File: contracts/contracts/rem.system/src/producer_pay.cpp [hunk 2]
```diff
@@ -182,6 +182,30 @@ namespace eosiosystem {
...
+            const auto producer_name = _gstate.last_schedule[producer_index].first;
+            const auto& prod = _producers.get(producer_name.value);
+            if( std::find(active_producers.begin(), active_producers.end(), producer_name) == active_producers.end() ) {
+              _producers.modify(prod, same_payer, [&](auto& p) {
+                 p.top21_chosen_time = time_point(eosio::seconds(0));
+              });
+            }
```
Comment (roman-tik) [inline contracts/contracts/rem.system/src/producer_pay.cpp]: lets's add `else` here and reduce `active_producers` vector, so the next `for` below iterates only over the newly added active producers without the need to do that `find_if`.

File: contracts/contracts/rem.system/src/rem.system.cpp [hunk 3]
```diff
@@ -107,6 +107,20 @@ namespace eosiosystem {
       _gremstate.per_vote_share = vote_share;
    }
 
+   void system_contract::setinacttime( uint64_t period_in_minutes ) {
+   require_auth(_self);
```
Comment (roman-tik) [inline contracts/contracts/rem.system/src/rem.system.cpp]:
```suggestion
   require_auth(get_self());
```

File: contracts/contracts/rem.system/src/rem.system.cpp [hunk 4]
```diff
@@ -107,6 +107,20 @@ namespace eosiosystem {
...
+
+   check(period_in_minutes != 0, "block producer maximum inactivity time cannot be zero");
+   _gremstate.producer_max_inactivity_time = eosio::minutes(period_in_minutes);
+    }
+
+    void system_contract::setpnshperiod( uint64_t period_in_days ) {
+       require_auth(_self);
```
Comment (roman-tik) [inline contracts/contracts/rem.system/src/rem.system.cpp]:
```suggestion
       require_auth(get_self());
```

File: contracts/contracts/rem.system/src/rem.system.cpp [hunk 5]
```diff
@@ -216,6 +230,25 @@ namespace eosiosystem {
       });
    }
 
+   void system_contract::punishprod( const name& producer ) {
+    auto prod = _producers.find( producer.value );
+    check( prod != _producers.end(), "producer not found" );
+
+    const auto ct = current_time_point();
+    check( prod->active() && prod->top21_chosen_time != time_point(eosio::seconds(0)), "can only punish top21 active producers" );
+    check( ct - _gremstate.producer_max_inactivity_time > prod->punished_until, "punishment period is not over yet" );
```
Comment (roman-tik) [inline contracts/contracts/rem.system/src/rem.system.cpp]: not sure I understood the purpose of this check here

**Review by roman-tik — CHANGES_REQUESTED** [review]
please also add tests

File: contracts/contracts/rem.system/src/rem.system.cpp (hunk 5 above)
Comment (Art17) [reply contracts/contracts/rem.system/src/rem.system.cpp]: Removed.

File: contracts/contracts/rem.system/src/rem.system.cpp (hunk 4 above)
Comment (Art17) [reply contracts/contracts/rem.system/src/rem.system.cpp]: Done.

File: contracts/contracts/rem.system/src/rem.system.cpp (hunk 3 above)
Comment (Art17) [reply contracts/contracts/rem.system/src/rem.system.cpp]: Done.

File: contracts/contracts/rem.system/src/producer_pay.cpp (hunk 2 above)
Comment (Art17) [reply contracts/contracts/rem.system/src/producer_pay.cpp]: What for?

File: contracts/contracts/rem.system/src/producer_pay.cpp (hunk 1 above)
Comment (Art17) [reply contracts/contracts/rem.system/src/producer_pay.cpp]: Done.

**Review by roman-tik — APPROVED** [review]

**Pull request merged**